### PR TITLE
 test: check git version

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

Added an esm ```js``` file to check git version for compatibility. If found not compatible it will prevent unit testing.

## Context:
PR created to solve the issue #11167 during tests only

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

I wanted to ask how workflow testing works, cause when I run the workflow then the tests on different OS run, and then lint runs separately, so how is the checking on all three OS happening, I do not get that.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -14,5 +14,31 @@
...
+    "statusBar.foreground": "#15202b",
+    "statusBarItem.hoverBackground": "#e6e6e6",
+    "tab.activeBorder": "#ffffff",
+    "titleBar.activeBackground": "#eeeeee",
+    "titleBar.activeForeground": "#15202b",
+    "titleBar.border": "#eeeeee",
+    "titleBar.inactiveBackground": "#eeeeee99",
+    "titleBar.inactiveForeground": "#15202b99"
+  },
+  "peacock.color": "#3a96ba"
```
Comment (rarkins) [inline .vscode/settings.json:43]: revert

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -5,6 +5,7 @@ on:
     branches:
       - main
       - v25
+      - feat/check-git-version
```
Comment (rarkins) [inline .github/workflows/build.yml:8]: revert before merge

File: tools/check-git-version.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const shell = require('shelljs');
+
+function checkGitVersion() {
+  try {
+    const gitVersion = shell
+      .exec('git --version', { silent: true })
```
Comment (rarkins) [inline tools/check-git-version.js:6]: Should we use `simple-git` for this because what simple-git uses is what is important?

File: tools/check-git-version.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      .stdout.toString()
+      .slice(12)
```
Comment (rarkins) [inline tools/check-git-version.js:8]: Prefer no magic numbers

File: tools/check-git-version.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      .split('.');
+    if (
+      parseInt(gitVersion[0], 10) >= 2 &&
+      parseInt(gitVersion[1], 10) >= 22 &&
```
Comment (rarkins) [inline tools/check-git-version.js:12]: What if it's version 3.0.0?

File: tools/check-git-version.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      parseInt(gitVersion[2], 10) >= 0
+    ) {
+      process.exit(0);
+    } else {
+      // shell.echo('WARNING: GIT VERSION NOT COMAPTIBLE');
+      throw new Error('GIT VERSION NOT COMAPTIBLE');
```
Comment (rarkins) [inline tools/check-git-version.js:18]:
```suggestion
      throw new Error('Minimum git version 2.33.0 is required');
```

File: tools/check-git-version.js (hunk 5 above)
Comment (rarkins) [reply tools/check-git-version.js:12]: Also, needs to be 2.33.0 of higher

**Review by rarkins — CHANGES_REQUESTED** [review]

File: tools/check-git-version.js (hunk 3 above)
Comment (RahulGautamSingh) [reply tools/check-git-version.js:6]: We will get the same o/p string while using shell/simple-git, I do not see any extra benefit it using Simple Git, sorry if  I am missing something..

File: tools/check-git-version.js (hunk 4 above)
Comment (RahulGautamSingh) [reply tools/check-git-version.js:8]: Can I use a regex to retrieve the version numbers from the string?
Thinking of using this one ```/[\d]+(?=.)/g```  it returns the major, minor, and revision version number in an array and then compares to GIT_MINIMUM_VERSION.

File: tools/check-git-version.js (hunk 3 above)
Comment (rarkins) [reply tools/check-git-version.js:6]: Please use simple-git then

File: tools/check-git-version.js (hunk 4 above)
Comment (rarkins) [reply tools/check-git-version.js:8]: It just needs to work reliably

File: tools/check-git-version.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      .slice(12)
+      .split('.');
+    if (
+      parseInt(gitVersion[0], 10) >= 2 &&
+      parseInt(gitVersion[1], 10) >= 22 &&
+      parseInt(gitVersion[2], 10) >= 0
+    ) {
+      process.exit(0);
+    } else {
+      throw new Error('Minimum git version 2.33.0 is required');
```
Comment (rarkins) [inline tools/check-git-version.js:17]:
```suggestion
      throw new Error(`Minimum git version ${minimumGitVersion} is required`);
```
Better to define the minimum version at the top of the file and then reuse it in strings/comparisons later

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by HonkingGoose** [comment]
Hi @RahulGautamSingh, I see you're hard at work on your first PR! 😄 ❤️ 

Did you know that you can close a issue automatically with your PR?

Quote from GitHub docs:

>To link a pull request to an issue to show that a fix is in progress and to automatically close the issue when someone merges the pull request, type one of the following keywords followed by a reference to the issue. For example, Closes `#10` or `Fixes octo-org/octo-repo#100`.

Read the [GitHub docs, using keywords in issues and pull requests](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) to learn more, and to see the full list of valid keywords.

---

Also the PR title needs a small fix, you need to put a space between `feat:` and `check git version` for it to pass the "Semantic Pull Request" status check.

File: tools/check-git-version.js (hunk 4 above)
Comment (viceice) [reply tools/check-git-version.js:8]: I think simple-git can do this parsing for us more easy. Have you tested it?

File: tools/check-git-version.js (hunk 3 above)
Comment (RahulGautamSingh) [reply tools/check-git-version.js:8]: Sure

File: tools/check-git-version.js (hunk 4 above)
Comment (RahulGautamSingh) [reply tools/check-git-version.js]: I could not find any command in simple-git to do the parsing for us.
Also, I was wondering if I can use the npm-package ```compare-versions``.

Comment (viceice) [reply tools/check-git-version.js]: You can use semver package. You first need to coerse, then you can do simple comparison.

Comment (viceice) [reply tools/check-git-version.js]: If simple-git didn't have a prepared function use shelljs. But better use a regex to extract the version.

Comment (RahulGautamSingh) [reply tools/check-git-version.js]: I did it using regex and semver.

File: tools/check-git-version.js (hunk 5 above)
Comment (RahulGautamSingh) [reply tools/check-git-version.js]: I added a GIT_MINIMUM_VESRION in the updated code and used it to compare to the user version we got..

File: tools/check-git-version.js [hunk 8]
```diff
@@ -10,7 +10,9 @@ function checkGitVersion() {
       .stdout.toString();
     gitVersion = gitVersion.match(regex)[0];
     if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
-      throw new Error('Minimum git version 2.33.0 is required');
+      throw new Error(
+        'Minimum git version ' + GIT_MINIMUM_VERSION + ' is required'
```
Comment (HonkingGoose) [inline tools/check-git-version.js:14]:
```suggestion
        'Minimum Git version ' + GIT_MINIMUM_VERSION + ' is required'
```

Git is a proper noun, should be capitalized. 😉

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -5,6 +5,7 @@ on:
     branches:
       - main
       - v25
+
```
Comment (viceice) [inline .github/workflows/build.yml:8]:
```suggestion
```

File: .vscode/extensions.json [hunk 10]
```diff
@@ -1,7 +0,0 @@
-{
```
Comment (viceice) [inline .vscode/extensions.json:1]: revert

File: .vscode/launch.json [hunk 11]
```diff
@@ -1,80 +0,0 @@
-{
```
Comment (viceice) [inline .vscode/launch.json:1]: revert

File: .vscode/settings.json [hunk 12]
```diff
@@ -1,18 +0,0 @@
-{
```
Comment (viceice) [inline .vscode/settings.json:1]: revert

File: tools/check-git-version.js (hunk 8 above)
Comment (viceice) [reply tools/check-git-version.js:14]:
```suggestion
        `Minimum git version ${GIT_MINIMUM_VERSION} is required`
```

Use javascript string templates

File: tools/check-git-version.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+const semver = require('semver');
+const shell = require('shelljs');
+
+const GIT_MINIMUM_VERSION = '2.33.0';
+function checkGitVersion() {
```
Comment (viceice) [inline tools/check-git-version.js:5]:
```suggestion
```

Don't use wrapper function, no benefit, only more complexity

File: tools/check-git-version.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+const semver = require('semver');
+const shell = require('shelljs');
```
Comment (viceice) [inline tools/check-git-version.js:2]: convert to esm imports

File: tools/check-git-version.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+const GIT_MINIMUM_VERSION = '2.33.0';
+function checkGitVersion() {
+  try {
+    const regex = /[\d.]+(?=)/g;
```
Comment (viceice) [inline tools/check-git-version.js:7]: move to module scop const
```suggestion
    const regex = /\d+\.\d+\.\d+/;
```

Only match version of:

```
git version 2.33.0
git version 2.33.0.windows.2
```

File: tools/check-git-version.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    let gitVersion = shell
+      .exec('git --version', { silent: true })
+      .stdout.toString();
+    gitVersion = gitVersion.match(regex)[0];
```
Comment (viceice) [inline tools/check-git-version.js:11]:
```suggestion
    const { stdout } = shell.exec('git --version', { silent: true });
    const [ gitVersion ] = regex.exec(stdout);
```

Simplify

File: tsconfig.json [hunk 17]
```diff
@@ -15,7 +15,7 @@
     "experimentalDecorators": true,
     "useUnknownInCatchVariables": false /* we aren't prepared for enabling this by default since ts 4.4*/,
     "lib": ["es2018"],
-    "types": ["node", "jest", "jest-extended"],
+    "types": ["jest", "jest-extended", "node"],
```
Comment (viceice) [inline tsconfig.json:18]: revert

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
This is not ready for review:
- Deletes three files it should not
- Does not use `simple-git`

File: tools/check-git-version.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+const GIT_MINIMUM_VERSION = '2.33.0';
+const git = simpleGit();
+async function checkGitVersion() {
+  try {
+    const regex = /\d+\.\d+\.\d+/;
+    const stdout = await git.raw('--version');
+    const [gitVersion] = regex.exec(stdout);
+    if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
+      throw new Error(`Minimum git version ${GIT_MINIMUM_VERSION} is required`);
```
Comment (HonkingGoose) [inline tools/check-git-version.js:13]:
```suggestion
      throw new Error(`Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
```

Git is a proper noun, should be capitalized. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: package.json [hunk 19]
```diff
@@ -279,5 +280,6 @@
   },
   "files": [
     "dist"
-  ]
+  ],
+  "type":"module"
```
Comment (rarkins) [inline package.json:284]: Is this change necessary?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: package.json (hunk 19 above)
Comment (viceice) [reply package.json:284]: no, revert this

File: tools/check-git-version.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    }
+    process.exit(0);
+  } catch (err) {
+    shell.echo('ERROR:', err.message);
+    process.exit(1);
+  }
+}
+checkGitVersion();
```
[the remark this answers is not in the document]
Comment (viceice) [reply tools/check-git-version.js:21]: nope

File: tools/check-git-version.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import semver from 'semver';
+import shell from 'shelljs';
+import simpleGit from 'simple-git';
+
+const GIT_MINIMUM_VERSION = '2.33.0';
+const git = simpleGit();
+async function checkGitVersion() {
```
Comment (viceice) [inline tools/check-git-version.js:7]:
```suggestion
// eslint-disable-next-line @typescript-eslint/no-floating-promises
(async () => {
```

Use simple arrow function with wrapper

File: tools/check-git-version.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+    if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
+      throw new Error(`Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
+    }
+    process.exit(0);
+  } catch (err) {
+    shell.echo('ERROR:', err.message);
+    process.exit(1);
+  }
+}
+checkGitVersion();
```
Comment (viceice) [inline tools/check-git-version.js:21]:
```suggestion
})();
```

File: tools/check-git-version.js [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+import simpleGit from 'simple-git';
+
+const GIT_MINIMUM_VERSION = '2.33.0';
+const git = simpleGit();
+async function checkGitVersion() {
+  try {
+    const regex = /\d+\.\d+\.\d+/;
+    const stdout = await git.raw('--version');
+    const [gitVersion] = regex.exec(stdout);
+    if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
```
[the remark this answers is not in the document]
Comment (viceice) [reply tools/check-git-version.js:12]: it will thow an error and catched below

File: package.json (hunk 19 above)
Comment (RahulGautamSingh) [reply package.json:284]: I had to do these to be able to use ```import``` keyword.

Comment (viceice) [reply package.json:284]: nope, there is a seperate `package.json` in tools directory, which already do this, works fine for the other tools

Comment (viceice) [reply package.json:284]: no, please rename file extension to `mjs` and it should work as is.

Comment (RahulGautamSingh) [reply package.json:284]: Sure thanks..I for some reason thought I should keep it ```js```

File: package.json [hunk 24]
```diff
@@ -19,10 +19,11 @@
     "eslint-fix": "eslint --ext .js,.mjs,.ts --fix lib/ test/ tools/",
     "generate": "run-s generate:*",
     "generate:imports": "node tools/generate-imports.mjs",
+    "git-check": "node tools/check-git-version.js",
```
Comment (HonkingGoose) [inline package.json:22]:
```suggestion
    "git-check": "node tools/check-git-version.mjs",
```

Not sure what the correct filename extension is when you run a script in `package.json`.

The file is called `check-git-version.mjs` in the PR, but the script uses `.js` as filename ending.

Comment (RahulGautamSingh) [reply package.json]: Comitted

**Comment by rarkins** [comment]
This needs to be tested against Windows and Mac VMs before we merge

**Review by rarkins — DISMISSED** [review]

File: tools/check-git-version.mjs [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
+(async () => {
+  try {
+    const regex = /\d+\.\d+\.\d+/;
+    const stdout = await git.raw('--version');
+    const [gitVersion] = regex.exec(stdout);
+    if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
+      throw new Error(`Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
+    }
+    console.log('Found git version: ' + gitVersion);
```
Comment (viceice) [inline tools/check-git-version.mjs:16]:
```suggestion
    shell.echo('Found git version: ', gitVersion);
```

Suppress eslint issue

**Review by viceice — APPROVED** [review]

File: tools/check-git-version.mjs [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+const GIT_MINIMUM_VERSION = '2.33.0';
+const git = simpleGit();
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
+(async () => {
+  try {
+    const regex = /\d+\.\d+\.\d+/;
+    const stdout = await git.raw('--version');
+    const [gitVersion] = regex.exec(stdout);
+    if (semver.lt(gitVersion, GIT_MINIMUM_VERSION)) {
+      throw new Error(`Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
```
Comment (viceice) [inline tools/check-git-version.mjs:14]:
```suggestion
      if (process.env.CI) {
        shell.echo(`::error ::Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
      } else {
        throw new Error(`Minimum Git version ${GIT_MINIMUM_VERSION} is required`);
      }
```

do not fail on CI

File: tools/check-git-version.mjs [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    shell.echo('Found git version: ', gitVersion);
+    process.exit(0);
+  } catch (err) {
+    shell.echo('ERROR:', err.message);
```
Comment (viceice) [inline tools/check-git-version.mjs:19]:
```suggestion
    shell.echo('::error ::', err.message);
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: tools/check-git-version.mjs (hunk 27 above)
Comment (viceice) [reply tools/check-git-version.mjs:19]: ignore

File: tools/check-git-version.mjs (hunk 26 above)
Comment (viceice) [reply tools/check-git-version.mjs:14]: https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message

**Review by viceice — DISMISSED** [review]

**Review by viceice — COMMENTED** [review]
needs prettier-fix run 🙃

**Comment by RahulGautamSingh** [comment]
> This needs to be tested against Windows and Mac VMs before we merge


> needs prettier-fix run 🙃

I will pull it and commit again now.

File: .vscode/settings.json [hunk 28]
```diff
@@ -14,5 +14,31 @@
...
+    "statusBar.foreground": "#15202b",
+    "statusBarItem.hoverBackground": "#e6e6e6",
+    "tab.activeBorder": "#ffffff",
+    "titleBar.activeBackground": "#eeeeee",
+    "titleBar.activeForeground": "#15202b",
+    "titleBar.border": "#eeeeee",
+    "titleBar.inactiveBackground": "#eeeeee99",
+    "titleBar.inactiveForeground": "#15202b99"
+  },
+  "peacock.color": "#e6e6e6"
```
Comment (HonkingGoose) [inline .vscode/settings.json:43]: You're pushing some color customizations, please undo. 😉

Comment (RahulGautamSingh) [reply .vscode/settings.json:43]: Vscode adds it automatically such a pain.😣

Comment (rarkins) [reply .vscode/settings.json:43]: @RahulGautamSingh please check your code before pushing things like this. You're wasting reviewer time by doing this repeatedly

**Comment by RahulGautamSingh** [comment]
Should I make it ready for review now,

**Review by viceice — APPROVED** [review]

File: .vscode/settings.json (hunk 28 above)
Comment (viceice) [reply .vscode/settings.json]: I'm using vscodew too and it's not changing any settings automatically. so please check your global settings / addons
